core/src/main/java/org/infinispan/interceptors/impl/BaseSequentialInvocationContext.java

```diff
@@ -0,0 +1,408 @@
+package org.infinispan.interceptors.impl;
+
+import org.infinispan.commands.VisitableCommand;
+import org.infinispan.commons.CacheException;
+import org.infinispan.context.InvocationContext;
+import org.infinispan.context.SequentialInvocationContext;
+import org.infinispan.interceptors.SequentialInterceptor;
+import org.infinispan.interceptors.base.CommandInterceptor;
+import org.infinispan.util.concurrent.CompletableFutures;
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+import org.jboss.logging.MDC;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+
+/**
+ * This base class implements the {@link org.infinispan.context.SequentialInvocationContext} methods.
+ *
+ * @author Dan Berindei
+ * @since 8.2
+ */
+public abstract class BaseSequentialInvocationContext
+      implements InvocationContext, SequentialInvocationContext {
+   private static final Log log = LogFactory.getLog(BaseSequentialInvocationContext.class);
+   private static final boolean trace = log.isTraceEnabled();
+   // Enable when debugging BaseSequentialInvocationContext itself
+   private static final boolean EXTRA_LOGS = SecurityActions.getBooleanProperty("org.infinispan.debug.BaseSequentialInvocationContext");
+
+   private static final CompletableFuture<Void> CONTINUE_INVOCATION = CompletableFuture.completedFuture(null);
+   private static final int INVOKE_NEXT = 0;
```

Comment:
Should we use an enum for these?  Then we can have switch statements instead of if else blocks